src/go/util/url_util.go

```diff
@@ -198,6 +198,54 @@ func WildcardMatcherForPath(uri string) string {
 	return "^" + matcher + "$"
 }
 
+// This function return the uri string with snakeNames replaced with jsonName.
+// It assume:
+//   - the input uri template is valid and it won't verify the uri.
+//   - each snakeName as variable in the input uri appear equal to or less than once.
+//
+// It uses the hacky substring replacement:
+//   - find the first appearance of snakeName, the char before which is '{' or '.',
+//     the char after which is '}' or '.' or '='
+//   - replace that substring with the jsonName
+//
+// It should match the variable name extraction behavior in
+// https://github.com/GoogleCloudPlatform/esp-v2/blob/34314a46a54001f83508071e78596cba08b6f456/src/api_proxy/path_matcher/http_template_test.cc
+//
+// TODO(taoxuy@): extract variable name by syntax parsing.
+func SnakeNamesToJsonNamesInPathParam(uri string, snakeNameToJsonName map[string]string) string {
+	findPathParamIndex := func(uri, snakeName string) int {
+		for {
+			index := strings.Index(uri, snakeName)
+			if index == -1 {
+				return -1
+			}
+
+			if index != 0 && (uri[index-1] == '{' || uri[index-1] == '.') && (uri[index+len(snakeName)] == '}' || uri[index+len(snakeName)] == '.' || uri[index+len(snakeName)] == '=') {
```

Comment:
This is very confusing to me. Can we replace with a regex capture group? Or can you add a comment on what's being checked?

Reply:
I described the matching logic in the line 207 and also commented.  Is it clear to you? 

The regex I think of is not supported in go lib(not sure if there is equivalent one). As it is not that complicated, we implement here.

Reply:
Thanks, makes sense